### PR TITLE
Remove Wayland theme intermediates

### PR DESCRIPTION
**Pull request by chrisduerr**
This removes the intermediate struct for passing a Wayland theme to
allow the user direct implementation of the trait.

By passing the trait directly, it is possible for downstream users to
have more freedom with customization without relying on winit to offer
these options as fields.

It should also make maintenance easier, since winit already doesn't
implement all the functions which are offered by the smithay client
toolkit.

At this point this is mainly a proposal to resolve the issue that it's not possible to set the (hovered) icon color for each icon, while Smithay allows for it. Alternative implementations could just add six more fields to the struct.

It would also be possible to offer both options, so winit could stick to the limited struct with their own theme implementation and at the same time allow specifying a custom theme implementation.

### CODE REVIEW

**Comment by vberger** [comment]
The main issue I have with exposing this trait is that this effectively makes SCTK a public dependency of winit, which I would advise against as I expect it to change quite a lot yet.

**Comment by chrisduerr** [comment]
> The main issue I have with exposing this trait is that this effectively makes SCTK a public dependency of winit, which I would advise against as I expect it to change quite a lot yet.

As long as breaking changes to the crate are marked properly by bumping the version accordingly, this shouldn't be too big of a deal, right?

I see the issue, however changes to this trait are one of the reasons why I'm proposing this. Alacritty is interested in the functionality while winit doesn't expose it yet, while winit itself does not really offer anything Alacritty is interested in with this feature. So that mainly makes things harder for downstream, while causing a maintenance burden for winit itself.

This issue could just be fixed since I don't see much changes to the trait that would be required in the future, however even then the intermediate struct seems more work than it's worth to me personally.

**Comment by JMurph2015** [comment]
I think my approach over at JMurph2015/winit was sufficient for what we needed to do.  It's not complete (notably I needed my upstream changes on sctk to land before pulling here), but it's a good start.

**Comment by chrisduerr** [comment]
@JMurph2015 That's precisely what I was trying to avoid. Implementing the changes wouldn't be a problem, but like this it just creates a bit of a mess really with one huge struct holding all kinds of variations.

**Comment by JMurph2015** [comment]
However, it's even more messy to expose the internal implementation details of winit.

**Comment by chrisduerr** [comment]
> However, it's even more messy to expose the internal implementation details of winit.

This is not an internal implementation detail of winit. This is a direct exposure of the API how it is meant to be used by clients. Winit just wraps an existing API and repackages it with a different API. It makes very little sense imho.

**Comment by chrisduerr** [comment]
I'd like to know how to move forward with this. I feel like all possible solutions would be trivial to implement and this is just waiting for a decision on the best possible choice.

It would be a shame if half the functionality of SCTK is never exposed in winit, simply because nobody could make a decision on how to move forward. This is also the cause for a PR sitting around for just two weeks shy of a year in the Alacritty repository. So I definitely feel like we should be ready to move forward with a solution for now.

I've proposed my suggestion with this PR and while @vberger has mentioned that SCTK as a public depedency might be cause for headaches, as a user I think it's the easiest way to make the features of SCTK actually usable downstream. This also only exposes a very small part of the API which isn't likely to change much itself.

So can we get a decision on how to move forward with this? There's not much that can be done without any suggested changes.

**Comment by vberger** [comment]
This specific part of SCTK's API is indeed unlikely to change very much in the future, though as SCTK's maintainer I cannot reasonably promise that it won't ever change.

I'm not going to veto this change nor approve it, as per #972 I'm trying to reduce my involvement in Winit, and this starts by no longer taking part into API design decisions.

**Comment by chrisduerr** [comment]
> though as SCTK's maintainer I cannot reasonably promise that it won't ever change.

Is there any need to do so? Chances that there's a non-breaking update extending its functionality to benefit are likely higher than the chances for a breaking change that would require a breaking winit bump. And breaking updates to this API that would require changes in winit are likely a breaking change anyways, it would just require additional patching in winit.

> I'm not going to veto this change nor approve it, as per #972 I'm trying to reduce my involvement in Winit, and this starts by no longer taking part into API design decisions.

Yeah, I understand that and that's a good idea, however I would still like a decision by someone who is actually still maintaining winit. Stalling progress on Wayland for years does not sound like a healthy path forwards.

**Comment by goddessfreya** [comment]
cc @murarth, what's your opinion?

@chrisduerr Is this the only place where we expose SCTK, or do we do so somewhere else? If this is the only place, won't Winit clients have to start synchronizing their version of SCTK with winit's? If this struct is unlikely to change, as you mentioned, why not amend winit's structure instead?

**Comment by murarth** [comment]
I don't know if this has been considered, but it would also be possible to copy the current definition of smithay's `Theme` trait into winit and implement `smithay::Theme` on an internal wrapper type containing a `winit::Theme` instance. This would give applications the flexibility of the trait API while allowing winit to control when breaking changes are introduced.

However, if that's too cumbersome (are applications already using smithay directly and implementing `Theme` on local types?), we could simply re-export the smithay trait and deal with any breakage when it occurs.

**Comment by chrisduerr** [comment]
> Is this the only place where we expose SCTK

As far as I'm aware, yes.

> If this struct is unlikely to change, as you mentioned, why not amend winit's structure instead?

I'm not sure if it is? I find winit's API more difficult to use and we'll have to make changes every time SCTK is altered. I find that to be a maintenance burden and I don't really want to send a winit PR every time things change in SCTK here.

> I don't know if this has been considered, but it would also be possible to copy the current definition of smithay's Theme trait into winit and implement smithay::Theme on an internal wrapper type containing a winit::Theme instance.

This would indeed be an option if we assume that the only problem is that the winit API is more annoying to use and a public SCTK dependency is undesirable.

However it would still mean that winit needs to rehash whatever SCTK does.

> However, if that's too cumbersome (are applications already using smithay directly and implementing Theme on local types?), we could simply re-export the smithay trait and deal with a major version change if it occurs.

If we're talking about winit users, I don't think there's any point to implement `Theme`, since winit's current API does not allow passing that.

**Comment by murarth** [comment]
> However it would still mean that winit needs to rehash whatever SCTK does.

Yes, if SCTK were to make a change to its `Theme` trait, winit would need to update its own `Theme` trait and publish a new release. However, given the simplicity of the `Theme` trait, I think that any such change would likely be trivial and require only a non-breaking winit release. I think this is a small maintenance cost to preserve winit's ability to manage its SCTK dependency version without requiring breaking releases.

**Comment by kchibisov** [comment]
I do understand the idea behind @chrisduerr solution here and it would sound about right, but there's one point I'd like to bring. If it happened that SCTK theme trait was changed it's likely some other chunk of SCTK APIs were also changed and you can't just bump SCTK to get new fixes, since it'll break winit. So if you're updating to a newer SCTK there's nothing wrong in updating to a new theme trait all together. Also applications could use a different version of SCTK mixed with winit, as it was already stated. So I think creating a wrapper around SCTK theme in winit is a necessary evil and is better for a clients, since they shouldn't even know what SCTK is.

**Comment by chrisduerr** [comment]
The Theme is now reimplemented, but since we can't just implement `SCTKTheme for WinitTheme` (we don't own SCTK theme), it's a bit ugly.

I'm not a huge fan of this, but it does work.

File: src/platform/unix.rs [hunk 1]
```diff
@@ -406,3 +339,91 @@ impl MonitorHandleExtUnix for MonitorHandle {
...
+    fn minimize_button_color(&self, status: ButtonState) -> [u8; 4];
+
+    /// Icon color for the minimize button, defaults to the secondary color.
+    #[allow(unused_variables)]
+    fn minimize_button_icon_color(&self, status: ButtonState) -> [u8; 4] {
+        self.secondary_color(true)
+    }
+}
+
+impl SCTKTheme for Box<dyn Theme + Send> {
```
[the remark this answers is not in the document]
Comment (chrisduerr) [reply src/platform/unix.rs]: We can, though we'd still need the `+ Send + 'static`. So either would work, the question is just if we want to have a `WinitTheme` struct or a `Box`.

File: src/platform/unix.rs [hunk 2]
```diff
@@ -220,7 +152,7 @@ pub trait WindowExtUnix {
     fn wayland_display(&self) -> Option<*mut raw::c_void>;
 
     /// Sets the color theme of the client side window decorations on wayland
-    fn set_wayland_theme(&self, theme: WaylandTheme);
+    fn set_wayland_theme<T: Theme + Send + 'static>(&self, theme: T);
```
[the remark this answers is not in the document]
Comment (chrisduerr) [reply src/platform/unix.rs]: I've just added it on later, so I added it where the error was. Adding it to the trait makes more sense.

File: src/platform/unix.rs [hunk 3]
```diff
@@ -406,3 +338,94 @@ impl MonitorHandleExtUnix for MonitorHandle {
         self.inner.native_identifier()
     }
 }
+
+/// Wrapper for implementing SCTK's theme trait.
+pub struct WaylandTheme<T: Theme>(T);
```
Comment (murarth) [inline src/platform/unix.rs]: I believe this should be private. Or `pub(crate)`, if necessary.

Comment (chrisduerr) [reply src/platform/unix.rs]: Yeah, there's absolutely no reason why this would be public, muscle memory just kicked in for me.

**Comment by murarth** [comment]
@chrisduerr Needs another `cargo fmt`.

**Comment by murarth** [comment]
@chrisduerr Oh, I almost forgot: This also needs a "**Breaking**" changelog entry.

**Pull request merged**